File: 985-bag-of-tokens/bag-of-tokens.scala
object Solution {
    def bagOfTokensScore(tokens: Array[Int], power: Int): Int = {
        var res = 0
        var score = 0
        var left = 0
        var right = tokens.length - 1
        val sortedTokens = tokens.sorted 
        var currentPower = power 
        while (left <= right) {
            if (sortedTokens(left) <= currentPower) {
                currentPower -= sortedTokens(left) 
                score += 1
                left += 1
                res = math.max(score, res)
            }
            else if (score > 0) {
                currentPower += sortedTokens(right)
                right -= 1
                score -= 1
            }
            else {
                return res
            }
        }
        res 
    }
}

File: 985-bag-of-tokens/bag-of-tokens.py
class Solution:
    def bagOfTokensScore(self, tokens: List[int], power: int) -> int:
        low = 0
        high = len(tokens) - 1
        res, score = 0, 0
        tokens.sort()

        while low <= high:
            if power >= tokens[low]:
                score += 1
                power -= tokens[low]
                low += 1
                res = max(res, score)
            elif score > 0:
                score -= 1
                power += tokens[high]
                high -= 1
            else:
                return res

        return res       